### PR TITLE
Dashboard sns

### LINKED ISSUE #64: Dashboard - link with SNS
- [ ] store email and name in database for pdf
- [ ] store email and tag choices in SNS 
- [ ] create topic for tag if not already there

As an engineer I need to adhere to the brief and provide the requested outputs for the project

### LINKED ISSUE #73: Fix queries to return correct data
Some queries appear to be returning incorrect data such as the number of sales per tag as well as per country. These queries need to be reviewed and fixed to return the correct number of sales.

As an engineer, I need to notice issues in data and correct them.

### PR DESCRIPTION
**Pull request by alina-101**
Linked information submitted in subscription form to subscribe to SNS topic. 
Added emails to the database for those who subscribe to the pdf
Fixed small query issues in `database.py` to not return duplicate values

closes #64 
closes #73

### CODE REVIEW

File: dashboard/database.py [hunk 1]
```diff
@@ -177,37 +170,32 @@ def get_all_tags(_conn: Connection) -> list:
 
 
 @st.cache_data(ttl="1hr")
-def get_sales_by_country(_conn: Connection, n: int = 5) -> pd.DataFrame:
+def get_sales_by_country(_conn: Connection):
```
Comment (e-lemma) [inline dashboard/database.py:173]: I think you mistakenly removed the expected return type hint here.

File: dashboard/Home.py [hunk 2]
```diff
@@ -7,29 +7,29 @@
...
-        "Choose which album you would like to see", album_titles)
+    # chosen_album = st.selectbox(
+    #     "Choose which album you would like to see", album_titles)
 
-    chosen_album_data = database.get_album_sales_by_album(conn, chosen_album)
-    chosen_album_data
-    st.altair_chart(charts.get_albums_sales_line_graph(chosen_album_data))
+    # chosen_album_data = database.get_album_sales_by_album(conn, chosen_album)
+    # chosen_album_data
+    # st.altair_chart(charts.get_albums_sales_line_graph(chosen_album_data))
```
Comment (e-lemma) [inline dashboard/Home.py:35]: Why is this code commented out?

File: dashboard/database.py [hunk 3]
```diff
@@ -131,25 +131,18 @@ def get_sales_by_tag(_conn: Connection, n: int = 5) -> pd.DataFrame:
...
+    FROM tag AS t
+    LEFT JOIN album_tag_assignment AS ata ON t.tag_id = ata.tag_id
+    LEFT JOIN album_purchase AS ap ON ata.album_id = ap.album_id
+    LEFT JOIN track_tag_assignment AS tta ON t.tag_id = tta.tag_id
+    LEFT JOIN track_purchase AS tp ON tta.track_id = tp.track_id
+    GROUP BY t.name
+    ORDER BY total_sales DESC
+    LIMIT %s
+    ;
+    """
```
Comment (e-lemma) [inline dashboard/database.py:145]: Why was this query changed?

File: dashboard/pages/subscriptions.py [hunk 4]
```diff
@@ -1,26 +1,49 @@
 """A dashboard page that lets you sign up for notifications."""
-
+from database import get_all_tags, get_connection
 import streamlit as st
+import boto3
+from os import environ as ENV
+from dotenv import load_dotenv
```
Comment (e-lemma) [inline dashboard/pages/subscriptions.py:6]: Standard libraries (os) should come first, then third party, local.

File: dashboard/pages/subscriptions.py [hunk 5]
```diff
@@ -1,26 +1,49 @@
...
 
 
-from database import get_all_tags, get_connection
+# Function to create a subscription
+def create_subscription(protocol, endpoint, topic_arn):
```
Comment (e-lemma) [inline dashboard/pages/subscriptions.py:10]: Docstring needed

File: dashboard/pages/subscriptions.py [hunk 6]
```diff
@@ -1,26 +1,49 @@
...
+    response = sns_client.subscribe(
+        TopicArn=topic_arn,
+        Protocol=protocol,
+        Endpoint=endpoint
+    )
+    return response
```
Comment (e-lemma) [inline dashboard/pages/subscriptions.py:16]: Could be simplified:

```suggestion
    return sns_client.subscribe(
        TopicArn=topic_arn,
        Protocol=protocol,
        Endpoint=endpoint
    )
```

File: dashboard/pages/subscriptions.py [hunk 7]
```diff
@@ -1,26 +1,49 @@
...
+
+
+def create_topic(topic_name: str):
+    response = sns_client.create_topic(
+        Name=f"c11-bandcamp-{topic_name}",
+    )
```
Comment (e-lemma) [inline dashboard/pages/subscriptions.py:22]: - Docstring needed
- What is the variable `response` being used here for?

File: dashboard/database.py (hunk 1 above)
Comment (alina-101) [reply dashboard/database.py:173]: yes, added back in now

File: dashboard/Home.py (hunk 2 above)
Comment (alina-101) [reply dashboard/Home.py:35]: Forgot to remove it, deleted now

File: dashboard/database.py (hunk 3 above)
Comment (alina-101) [reply dashboard/database.py:145]: The first query was returning incorrect data. Zhi spotted it and gave me the correct query. It uses `DISTINCT`, `LEFT JOIN` and `ON` instead

Comment (e-lemma) [reply dashboard/database.py:145]: I see, good change!

File: dashboard/pages/subscriptions.py (hunk 4 above)
Comment (alina-101) [reply dashboard/pages/subscriptions.py:6]: good spot, fixed

File: dashboard/pages/subscriptions.py (hunk 7 above)
Comment (alina-101) [reply dashboard/pages/subscriptions.py:22]: It's not being used, removed it now

File: dashboard/pages/subscriptions.py [hunk 8]
```diff
@@ -1,27 +1,35 @@
...
 from dotenv import load_dotenv
+from database import get_all_tags, get_connection
 
 
-# Function to create a subscription
-def create_subscription(protocol, endpoint, topic_arn):
-    response = sns_client.subscribe(
-        TopicArn=topic_arn,
+def create_subscription(protocol, endpoint, arn):
+    """adds a subscription to a topic"""
```
Comment (e-lemma) [inline dashboard/pages/subscriptions.py:11]:
```suggestion
    """Adds a subscription to a topic"""
```

File: dashboard/database.py [hunk 9]
```diff
@@ -170,7 +170,7 @@ def get_all_tags(_conn: Connection) -> list:
 
 
 @st.cache_data(ttl="1hr")
-def get_sales_by_country(_conn: Connection):
+def get_sales_by_country(_conn: Connection) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (alina-101) [reply dashboard/database.py]: Whoops good spot, that should be `list(dict)`

File: dashboard/pages/subscriptions.py [hunk 10]
```diff
@@ -1,27 +1,35 @@
...
+        TopicArn=arn,
         Protocol=protocol,
         Endpoint=endpoint
     )
-    return response
 
 
 def create_topic(topic_name: str):
-    response = sns_client.create_topic(
+    """creates an sns topic"""
```
Comment (e-lemma) [inline dashboard/pages/subscriptions.py:20]:
```suggestion
    """Creates an sns topic"""
```

File: dashboard/pages/subscriptions.py [hunk 11]
```diff
@@ -1,27 +1,35 @@
...
+    sns_client.create_topic(
         Name=f"c11-bandcamp-{topic_name}",
     )
 
 
+def get_topics(client):
+    """returns a list of all sns topics"""
```
Comment (e-lemma) [inline dashboard/pages/subscriptions.py:27]:
```suggestion
    """Returns a list of all sns topics"""
```

**Review by e-lemma — APPROVED** [review]
Comments have been addressed, good to merge.

**Pull request merged**